File: packages/backend-python/toon_backend/routes/websocket.py
"""
WebSocket route
"""
import asyncio
import json
import time
from fastapi import APIRouter, WebSocket, WebSocketDisconnect
from toon_backend.event_emitter import EventEmitter, EventType, AgentEvent


def register_websocket_route(router: APIRouter, event_emitter: EventEmitter):
    """Register WebSocket route"""
    @router.websocket("/ws")
    async def websocket_endpoint(websocket: WebSocket):
        await websocket.accept()
        
        # Send connection confirmation
        await websocket.send_json({
            "type": "connected",
            "timestamp": int(time.time() * 1000)
        })
        
        # Set up event handler
        def handler(event: AgentEvent):
            asyncio.create_task(websocket.send_json(event.to_dict()))
        
        # Listen to all event types
        for event_type in EventType:
            event_emitter.on(event_type, handler)
        
        try:
            while True:
                # Receive messages from client
                data = await websocket.receive_json()
                
                # Handle client messages (e.g., subscribe/unsubscribe)
                if data.get("type") == "subscribe":
                    # Subscribe to specific event types
                    # Implementation can be extended
                    pass
        except WebSocketDisconnect:
            # Clean up handlers on disconnect
            for event_type in EventType:
                event_emitter.off(event_type, handler)

File: packages/backend-node/src/routes/websocket.ts
import { FastifyInstance } from 'fastify';
import websocket from '@fastify/websocket';
import { AgentEvent, EventType } from '@programsmagic/toon-core';
import { EventEmitter } from '../event-emitter.js';
import type { SocketStream } from '@fastify/websocket';

/**
 * Register WebSocket route
 */
export async function registerWebSocketRoute(
  fastify: FastifyInstance,
  eventEmitter: EventEmitter
): Promise<void> {
  await fastify.register(websocket);

  fastify.get('/ws', { websocket: true }, (connection: SocketStream) => {
    // Send connection confirmation
    connection.socket.send(
      JSON.stringify({
        type: 'connected',
        timestamp: Date.now(),
      })
    );

    // Set up event listener
    const handler = (event: AgentEvent) => {
      try {
        if (connection.socket.readyState === 1) { // WebSocket.OPEN
          connection.socket.send(JSON.stringify(event));
        }
      } catch (error) {
        // Client disconnected
        eventEmitter.off(EventType.HEARTBEAT, handler);
      }
    };

    // Listen to all event types
    Object.values(EventType).forEach((eventType) => {
      eventEmitter.on(eventType, handler);
    });

    // Handle incoming messages from client
    connection.socket.on('message', (message: Buffer) => {
      try {
        const data = JSON.parse(message.toString());
        // Handle client messages (e.g., subscribe/unsubscribe to specific events)
        if (data.type === 'subscribe') {
          // Subscribe to specific event types
          // Implementation can be extended
        }
      } catch (error) {
        // Invalid message format
      }
    });

    // Clean up on disconnect
    connection.socket.on('close', () => {
      Object.values(EventType).forEach((eventType) => {
        eventEmitter.off(eventType, handler);
      });
    });
  });
}

